File: app/src/main/java/com/magenta/rx/java/presenter/RetrofitPresenter.java
package com.magenta.rx.java.presenter;

import android.util.Log;

import com.magenta.rx.kotlin.event.TranslateAnswerEvent;
import com.magenta.rx.kotlin.loader.TranslateLoader;
import com.magenta.rx.kotlin.record.TranslateAnswer;

import org.greenrobot.eventbus.EventBus;

import javax.inject.Inject;

import rx.Observer;
import rx.schedulers.Schedulers;

public class RetrofitPresenter {

    private final TranslateLoader loader;

    @Inject
    public RetrofitPresenter(TranslateLoader loader) {
        this.loader = loader;
    }

    public void onLoadClick(final String text) {
        loader.load(text)
                .subscribeOn(Schedulers.io())
                .subscribe(new Observer<TranslateAnswer>() {
                    public void onCompleted() {
                        Log.d(getClass().getName(), "[onCompleted]");
                    }

                    public void onError(Throwable e) {
                        Log.e(getClass().getName(), "[onError]", e);
                    }

                    public void onNext(TranslateAnswer translate) {
                        Log.d(getClass().getName(), "[onNext]");
                        onReceiveAnswer(text, join(translate.getText(), ","));
                    }
                });
    }

    private String join(String[] text, String delimiter) {
        if (text != null && text.length > 0) {
            StringBuilder sb = new StringBuilder(text[0]);
            for (int i = 1; i < text.length; i++) {
                sb.append(delimiter).append(text[i]);
            }
            return sb.toString();
        }
        return "not found";
    }

    private void onReceiveAnswer(String text, String translate) {
        EventBus.getDefault().postSticky(new TranslateAnswerEvent(text, translate));
    }
}

File: app/src/main/kotlin/com/magenta/rx/kotlin/presenter/RetrofitPresenter.kt
package com.magenta.rx.kotlin.presenter

import android.text.TextUtils
import android.util.Log
import com.magenta.rx.kotlin.event.TranslateAnswerEvent
import com.magenta.rx.kotlin.loader.TranslateLoader
import com.magenta.rx.kotlin.record.TranslateAnswer
import org.greenrobot.eventbus.EventBus
import rx.Observer
import rx.schedulers.Schedulers
import javax.inject.Inject

class RetrofitPresenter @Inject constructor(private val loader: TranslateLoader) {

    fun onLoadClick(text: String) {
        loader.load(text)
                .subscribeOn(Schedulers.io())
                .subscribe(object : Observer<TranslateAnswer> {
                    override fun onCompleted() {
                        Log.d(javaClass.name, "[onCompleted]")
                    }

                    override fun onError(e: Throwable) {
                        Log.e(javaClass.name, "[onError]", e)
                    }

                    override fun onNext(translate: TranslateAnswer) {
                        Log.d(javaClass.name, "[onNext]")
                        onReceiveAnswer(text, TextUtils.join(",", translate.text))
                    }
                })
    }

    private fun join(text: Array<String>?, delimiter: String): String {
        if (text != null && text.isNotEmpty()) {
            val sb = StringBuilder(text[0])
            (1..text.size - 1).forEach { sb.append(delimiter).append(text[it]) }
            return sb.toString()
        }
        return "not found"
    }

    private fun onReceiveAnswer(text: String, translate: String) {
        EventBus.getDefault().postSticky(TranslateAnswerEvent(text, translate))
    }
}